lectures/5-Tutorials/SymmetricTensors.md

```diff
@@ -0,0 +1,793 @@
+```{math}
+U = \sum_i N_i
+```
+This invariance corresponds to a $\mathrm{U}(1)$ particle number symmetry, which can again be manifestly imposed when constructing the Hamiltonian terms as tensor maps. Recall that the irreps of $\mathrm{U}(1)$ are labeled by integers, and that the fusion of two irreps is given by addition.
+
+
+### Directly Constructing the Hamiltonian Terms
+
+We recall from our discussion on the $\mathbb{Z}_2$ symmetric Ising model that, in order to construct the Hamiltonian terms as symmetric tensors, we should work in the irrep basis where the symmetry transformation is block diagonal. In the current case, the symmetry operation is the particle number operator, which is already diagonal in the occupation number basis. Therefore, we don't need an additional local basis transformation this time, and can just observe that each local basis state can be identified with the $\mathrm{U}(1)$ irrep associated to the corresponding occupation number.
+
+Following the same approach as before, we first write down the action of the Hamiltonian terms in the irrep basis:
+
+```{math}
+\begin{align*}
+a_i^+ a_j^- \ket{n_i, n_j} &= \sqrt{(n_i + 1)n_j} \ket{n_i + 1, n_j - 1} \\
+a_i^- a_j^+ \ket{n_i, n_j} &= \sqrt{n_i(n_j + 1)} \ket{n_i - 1, n_j + 1} \\
+N \ket{n} &= n \ket{n}
+\end{align*}
+```
+
+It is then a simple observation that these matrix elements are exactly captured by the following $\mathrm{U}(1)$ fusion trees with corresponding block values:
+
+```{figure} ../_static/SymmetricTensors/U1_fusiontrees.svg
+:scale: 12%
+:name: U1_fusiontrees
+```
+
+This gives us all the information necessary to construct the corresponding `TensorMap`s. We follow the same steps as outlined in the previous example, starting with the construction of the physical space. This will now be a $\mathrm{U}(1)$ graded vector space `U1Space`, where each basis state $\ket{n}$ in the occupation number basis is represented by the corresponding $\mathrm{U}(1)$ irrep `U1Irrep(n)` with degeneracy 1. While this physical space is in principle infinite dimensional, we will impose a cutoff in occupation number at a maximum of 5 bosons per site, giving a 6-dimensional vector space:
+
+```{code-cell} julia
+cutoff = 5
+V = U1Space(n => 1 for n in 0:cutoff)
+```
+
+We can now initialize the $a^+ a^-$, $a^- a^+$ and $N$ operators as empty `TensorMap`s with the correct domain and codomain vector spaces, and fill in the nonzero blocks associated to [the fusion trees shown above](U1_fusiontrees). To do this we need access to the integer label of the $\mathrm{U}(1)$ irreps in the fusion and splitting trees, which can be accessed through the `charge` field of the `U1Irrep` type.
+
+```{code-cell} julia
+:tags: [hide-output]
+a⁺a⁻ = TensorMap(zeros, ComplexF64, V ⊗ V ← V ⊗ V)
+for (s, f) in fusiontrees(a⁺a⁻)
+    if s.uncoupled[1].charge == f.uncoupled[1].charge + 1 && s.uncoupled[2].charge == f.uncoupled[2].charge - 1
```

Comment:
Since it is not particularly nice that you need to access the inner `charge` field, I guess you could also write this as
`s.uncoupled[1] == only(f.uncoupled[1] ⊗ U1Irrep(1)) && s.uncoupled[2] == only(f.uncoupled[2] ⊗ U1Irrep(-1))` and similarly, in line with the example above. However, since also the matrix element depends on the `charge` field and I don't see a way around this, this is probably fine.

Reply:
Checking the constraints on the uncoupled and coupled irreps using the fields is indeed not very pretty, I'll change it to your suggestion.

Since accessing the fields of a `Sector` subtype instance is basically always necessary to set the matrix elements, I think it would be good to explicitly document the field names of all sectors defined in TensorKit.jl. I'll file a PR for that soon.